File: Ch01/Ex1_07.hs
{-
 - Ex1_07.hs
 - Kevin C. Baird
 - SICP in Haskell
 -
 - Redefine the tolerance to be a certain fraction 
 - of the guess, rather than a constant value.
 -}

module NewtonSquare where

sqrt :: Float -> Float
sqrt = sqrtIter 1.0

improve :: Float -> Float -> Float
improve guess x = (guess + (x/guess)) / 2.0

goodEnough :: Float -> Float -> Bool
goodEnough guess x = 
  (abs ((square guess) - x)) < tolerance guess

square :: Float -> Float
square x = x * x

sqrtIter :: Float -> Float -> Float
sqrtIter guess x
  | goodEnough guess x = guess
  | otherwise          = sqrtIter (improve guess x) x

tolerance :: Float -> Float
tolerance = (*) 0.001

File: Ch01/Ex1_07.scala
/**
  * Ex1_07.scala
  * Kevin C. Baird
  * SICP in Scala (http://www.scala-lang.org/documentation/)
  *
  * Redefine the tolerance to be a certain fraction
  * of the guess, rather than a constant value.
  */

object NewtonSquare extends App {
  def sqrt(x: Double) = {
    def sqrtIter(guess: Double, x: Double): Double = {
      def improve(guess: Double, x: Double) = {
        (guess + (x / guess)) / 2.0
      }
      def goodEnough(guess: Double, x: Double) = {
        def square(x: Double)    = { x * x }
        def tolerance(x: Double) = { x * 0.001 } /* TODO: figure out point-free */
        (math.abs((square(guess)) - x)) < tolerance(guess)
      }
      goodEnough(guess, x) match {
        case true  => guess
        case false => sqrtIter(improve(guess, x), x)
      }
    }
    sqrtIter(1.0, x) /* TODO: figure out point-free */
  }
}

/**
  * execute with
  * scala -i Ex1_07.scala -e "println(NewtonSquare.sqrt(2.0))"
  * until I figure out a better way
  */

